app/lib/company_shareholders_updator.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class CompanyShareholdersUpdator < ApplicationService
+  def initialize
+    @gpw_client = GPW::Client.new
+  end
+
+  def call(isin)
+    company = Company.find_by(isin: isin)
+    company_shareholders = @gpw_client.fetch_company_shareholders(company.stock, isin)
+    update_company_shareholders(company_shareholders, isin)
+  end
+
+  private
+
+  def update_company_shareholders(company_shareholders, isin)
+    company = Company.find_by(isin: isin)
+    company_shareholders.each do |shareholder_attrs|
+      shareholder = company.shareholders.find_or_create_by!(name: shareholder_attrs.fetch(:name))
+      shares_count = shareholder_attrs.fetch(:shares_count)
+      update_shares_count(company, shareholder, shares_count)
+    end
+  end
+
+  def update_shares_count(company, shareholder, shares_count)
+    share = company.shares.find_by(shareholder_id: shareholder.id)
+    share.update!(shares_count: shares_count)
+    share.touch
```

Comment:
I believe it will send 2 SQL write queries

you should be able to do it as an atomic operation